sys/xtimer: allow frequencies multiple of 921600Hz

kYc0o opened the pull request:
This PR will add xtimer compatibility for boards running at frequencies multiple of 576 (e.g. waspmote-pro running at 14.7456 MHz with timers prescaled to this frequency).

Partially (re) fixes #5597. An upcoming PR will fix the xtimer configuration on the waspmote.

kaspar030 commented on sys/include/div.h:
```diff
@@ -33,17 +33,52 @@ extern "C" {
+/**
+ * @brief Operand for 576/625
+ */
+#define MUL_576                576ul
```
pls drop these defines, they don't make sense.

kaspar030 commented on sys/include/div.h:
```diff
@@ -33,17 +33,52 @@ extern "C" {
+/**
+ * @brief Operand for 625/576
+ */
+#define MUL_625                625ul
```
here too

Kijewski replied:
I wonder what the correct value of `MUL_4711` would be. ;)

kYc0o replied:
It was to avoid the "ul" everywhere (8 and 16 bit requirement), but I can remove it.

kaspar030 replied:
for 625 you don't need UL. That only matters on the left-hand side of an expression (left-hand side defines type being used for calculation), or if the value is >16bit.

kYc0o replied:
OK, got it.

kYc0o replied:
Addressed. Thanks for the explanation.

kaspar030 commented on sys/include/div.h:
```diff
@@ -128,6 +196,73 @@ static inline uint64_t div_u64_by_15625div512(uint64_t val)
+static inline uint64_t div_u64_by_625div576(uint64_t val)
+{
+    /*
+     * This saves around 1400 bytes of ROM on Cortex-M platforms (both ARMv6 and
```
did you measure this? or C&P?

kYc0o replied:
Oops, copy-paste, I'll check if it does but I think it makes exactly the same as before. Moreover, it makes sense to take care of the overflow, I can just remove the comment but the code seems ok for me.

kaspar030 commented on sys/include/div.h:
```diff
@@ -128,6 +196,73 @@ static inline uint64_t div_u64_by_15625div512(uint64_t val)
+static inline uint32_t div_u64_by_576div625(uint64_t val)
+{
+    /*
+     * This saves around 1400 bytes of ROM on Cortex-M platforms (both ARMv6 and
```
same

kYc0o commented:
Ready?

lebrush commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -29,7 +29,78 @@ extern "C" {
+ */
+#if (XTIMER_HZ % 576 == 0)
 #if (XTIMER_SHIFT != 0)
+#if (XTIMER_HZ % 576 != 0)
```
You will never run into this, since you first checked for the opposite, right?
```C
#if (XTIMER_HZ % 576 == 0)
#if (XTIMER_HZ % 576 != 0)
```

kYc0o replied:
Yes, I can remove it but I need a way to throw the error if shifting is enabled.

lebrush commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -29,7 +29,78 @@ extern "C" {
+inline static uint64_t _xtimer_usec_from_ticks64(uint64_t ticks) {
+    return div_u64_by_576div625(ticks) >> XTIMER_SHIFT;
+}
+#else /* (XTIMER_HZ < 921600ul) */
```
`/* (XTIMER_HZ <= 921600ul) */`

lebrush commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -29,7 +29,78 @@ extern "C" {
+inline static uint64_t _xtimer_usec_from_ticks64(uint64_t ticks) {
+    return div_u64_by_576div625(ticks) << XTIMER_SHIFT;
+}
+#endif /* defined(XTIMER_SHIFT) && (XTIMER_SHIFT != 0) */
```
This is actually `/* (XTIMER_HZ > 921600ul) */` or `/* (XTIMER_HZ <= 921600ul) */` (depends on the notation you are using)

kYc0o replied:
Rather the first, and also `XTIMER_SHIFT`. Will fix.

lebrush commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -29,7 +29,78 @@ extern "C" {
+    return div_u64_by_576div625(ticks) << XTIMER_SHIFT;
+}
+#endif /* defined(XTIMER_SHIFT) && (XTIMER_SHIFT != 0) */
+#elif (XTIMER_HZ == 921600ul)
```
Is here expected to be: `(XTIMER_SHIFT == 0) && (XTIMER_HZ == 921600ul)`? If that's the case then is correct

kYc0o replied:
Yes, that's what is expected. I tried to reproduce a similar behaviour than for the 1MHz case.

kYc0o replied:
Yes that's what's expected.

lebrush commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -29,7 +29,78 @@ extern "C" {
+    return div_u64_by_576div625(ticks);
+}
+#endif /* (XTIMER_HZ == 921600ul) */
+#elif (XTIMER_SHIFT != 0) /* (XTIMER_HZ % 576 == 0) */
```
I think this branch belongs to the `#if (XTIMER_HZ % 576 == 0)`... then the code will enter the first time you test for `#if (XTIMER_SHIFT != 0)` and not here.

kYc0o replied:
Yes I was trying to solve the issue. Actually I need a way on which we can detect, first, if we're using either 1MHz or 921600Hz clock based timer, then we need to know if the shifting is used, to finally choose the right functions.

The current code will check first if there's shifting activated, then if it's a 1MHz or 921600Hz based clock. The process is repeated to ensure that we can choose 1MHz clocks.

kYc0o replied:
The code will enter here if the clock is not a multiple of 576, for which I need to check again if shifting is used.

kYc0o commented:
@lebrush I added more comments to clarify all those definitions. I still have no idea on how I can simplify them, so if you have suggestions they're welcomed (maybe I'm doing redundant things and I cannot see them by myself).

kYc0o commented:
Some thoughts @kaspar030 ?

kYc0o commented:
ping @kaspar030 !

smlng commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -29,16 +29,142 @@ extern "C" {
+ * If it uses shifting, just check if it's not multiple of 576, if so,
+ * throw an error
+ */
+#if (XTIMER_HZ % 576 != 0)
```
This will never be true, as you check above the opposite, or do I miss something?

kYc0o replied:
Yes I know, but then how to check this? I need to know if XTIMER_HZ is valid when using shifting, but I'll just remove this line...

kYc0o commented:
@lebrush comments addressed.

lebrush commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -29,16 +29,135 @@ extern "C" {
+ * Check for the actual frequency without any shifting (921600Hz)
+ * (XTIMER_SHIFT == 0)
+ */
+#elif (XTIMER_HZ == 921600ul)
```
I don't think you need a special case for that, shifting to any side by 0 (`XTIMER_SHIFT` should be 0 here). Would give the same result.

kYc0o replied:
Yes you're right, I think I can do the same for the 1MHz frequency.

kYc0o commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -29,16 +29,135 @@ extern "C" {
+/*
+ * If it's not greater, check if it's lower (XTIMER_HZ < 921600ul).
+ */
+#else
```
Ok I'll fix.

kYc0o commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -29,16 +29,135 @@ extern "C" {
+ * No more shifted multiple of 921600Hz frequencies to check
+ * (XTIMER_SHIFT != 0)
+ */
+#endif
```
Ok no problem, I just wanted to be as clear as possible, even for me. I find this code a bit complicated to understand in itself, so a bit more of documentation won't hurt IMHO.

kYc0o commented:
I somehow like your suggestion with the macros magic. However, we have an "unofficial" preference to keep macro usage as low as possible, thus very often we use static inline functions instead of macros.

I can try to adapt your proposition with static inline functions to make it "standard", but I'd like to have @gebart opinion on this, since he was the contributor of this code.

kaspar030 commented:
> One possibility is to use some macro magic

I'd strongly oppose this. Please rather try to find a solution without any macros. Most of the checks could be done in C-Code, being syntax checked, and optimized out by the compiler.

kYc0o commented:
>>One possibility is to use some macro magic
>I'd strongly oppose this. Please rather try to find a solution without any macros. Most of the checks could be done in C-Code, being syntax checked, and optimized out by the compiler.

That's what I meant when I said I could try to adapt those macros to static in line functions.

gebart commented:
@kYc0o feel free to go ahead with the macro magic -> static inline conversion.

lebrush commented:
I agree that macro magic makes sometimes code difficult to follow :-)
Another possibility is to create a separate `tick_conversion.h` for each frequency block (i.e. `tick_conversion_921600.h`) and include them in the corresponding `#if #else` blocks.

kYc0o commented:
I was trying to convert the ifdefs to inline functions but at the end I was adding more complexity, when why that code is there is to optimise the conversions.

I removed some unnecessary checks, I find now it's a bit more readable and works correctly.

I tested with samr21-xpro and arduino-mega2560 with success. What is missing is a test for cc2538 but xtimer is broken for that platform due to different reasons.

kYc0o commented:
For the separate file, IMHO, maybe it would be required if the cases of different frequencies increase. For now I think it's manageable, even if it was hard to get it in a less confusing way.

kYc0o commented:
Can I squash to get the CI working?

kYc0o commented:
ping @lebrush @kaspar030

lebrush commented:
Please squash

smlng commented on sys/include/div.h:
```diff
@@ -33,17 +33,42 @@ extern "C" {
  * @brief Approximation of (2**l)/d for d=15625, l=12, 32 bits
  */
-#define DIV_H_INV_15625_32    0x431bde83ul
+#define DIV_H_INV_15625_32     0x431bde83ul
```
suggestion put braces and capitalize constant type specification, i.e.
```
#define DIV_H_INV_15625_32     (0x431bde83UL)
```

IMHO using uppercase reads better, especially when defining HEX values.

gebart replied:
It doesn't matter if you have parentheses or not here, these are just magic numbers and never any complex expressions. As for the letter case, I don't care either way, I tend to prefer lowercase for the suffixes, but I don't see any point in forcing it.

smlng commented on sys/include/div.h:
```diff
@@ -75,6 +100,20 @@ static inline uint64_t div_u64_by_15625(uint64_t val)
+ */
+static inline uint64_t div_u64_by_625(uint64_t val)
+{
+    if (val > 10485759996ull) {
```
should/could this number also be defined above and use macro here? maybe short doc on where it comes from and why its used here might be good (for no so involved readers), too

kYc0o replied:
If you look at the code for the other calculation (div/15625) this kind of constant already exists. I know that this is not an "excuse" to do the same, but it was done this way to "respect" previous style.

I'll try to document both.

smlng replied:
I think its good to give a little insight/rational on why numbers that look kind of random are used here and there.

smlng replied:
and further this number is used at least twice here, so a macro might also help to reduce copy-paste errors

smlng commented on sys/include/div.h:
```diff
@@ -128,6 +186,65 @@ static inline uint64_t div_u64_by_15625div512(uint64_t val)
+ */
+static inline uint64_t div_u64_by_625div576(uint64_t val)
+{
+    if (val > 10485759996ull) {
```
ah the magic number again, used twice already 😉

smlng commented on sys/include/div.h:
```diff
@@ -128,6 +186,65 @@ static inline uint64_t div_u64_by_15625div512(uint64_t val)
+ * @param[in]   val     dividend
+ * @return      (val / (576/625))
+ */
+static inline uint32_t div_u64_by_576div625(uint64_t val)
```
look inconsistent, why is it returning a `uint32_t` while all the other (similar) functions `div_u64_by_xyz(...)` return `uint64_t`.

kYc0o replied:
Yes that's a mistake, will fix.

smlng commented on sys/include/div.h:
```diff
@@ -128,6 +186,65 @@ static inline uint64_t div_u64_by_15625div512(uint64_t val)
+ */
+static inline uint32_t div_u64_by_576div625(uint64_t val)
+{
+    if (val > 2473900978176ull) {
```
magic number again, without comment

smlng commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -29,16 +29,132 @@ extern "C" {
+}
+
+inline static uint64_t _xtimer_usec_from_ticks64(uint64_t ticks) {
+    return div_u64_by_576div625(ticks) >> XTIMER_SHIFT;
```
this return `uint32_t`, see above

kYc0o commented on sys/include/div.h:
```diff
@@ -33,17 +33,42 @@ extern "C" {
  * @brief Approximation of (2**l)/d for d=15625, l=12, 32 bits
  */
-#define DIV_H_INV_15625_32    0x431bde83ul
+#define DIV_H_INV_15625_32     0x431bde83ul
```
OK but that means also to change previous code by @gebart. I suppose I should change both?

smlng replied:
ah haven't looked into the existing code. mhm then I think its better to leave as is and fix that in a separate PR for your and existing code parts. I would rather not recommend that you fix existing stuff in this PR

kYc0o commented:
> see earlier comments by me too. Also note #6702, where I introduce automatic calc of XTIMER_SHIFT, which is only implemented for 1MHz base frequency

Yes I see, so would you mind to include this frequency in your PR? After addressing the comments I think this can be ready, but I won't start a race of merge ;-)

smlng commented:
@kYc0o I could wait with #6702 and include this one then, I'll mark it as WIP

kYc0o commented:
Comments addressed.

kYc0o commented:
Ping @kaspar030 @smlng

smlng commented:
@kYc0o ACK from my side, but lets keep in mind that xtimer could benefit from some improvements in code quality and structure for better readability.

smlng commented:
ah and don't forget to squash, as needed

kYc0o commented:
Super! Thanks. Yes I know this can maybe be rewritten in a more readable way, the current state has some logic and style, so maybe a general rework in the future can improve it.

kYc0o commented:
Well, everything seems good here, just a review from @kaspar030 is missing...

kYc0o commented:
ping @kaspar030

kYc0o commented:
re-ping @kaspar030

kYc0o commented:
@kaspar030 any objections to get this into the release? I have 2 more PRs which are based on this one, so I think they won't make it for the release if we don't merge this first soon.

smlng commented:
@kaspar030 could you please spare some time to ACK (or/and comment on) this and approve changes done?! I'd rather not dismiss your review, but have your trusted ACK here 😉

kYc0o commented:
@kaspar030 will you leave this PR open until the xtimer rework is done? in such case we can close it since it won't certainly be compatible with the new xtimer code.

kaspar030 commented:
> @kaspar030 will you leave this PR open until the xtimer rework is done? in such case we can close it since it won't certainly be compatible with the new xtimer code.

Hm, good question. ;)

By using a 32khz as backend for xtimer (not new in this PR), all code throughout the codebase would need to be checked if it's still working. E.g., are usec values ```<1/2**15``` rounded down to 0?

We need multiple timers...

kYc0o commented:
> E.g., are usec values <1/2**15 rounded down to 0?

Well, it will depend how _much_ usec you need, since [here](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/div.h#L69) we're certainly losing accuracy by shifting.

kYc0o commented:
What do we do with this one? It's needed to support xtimer on waspmote-pro...

gebart commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -29,16 +29,132 @@ extern "C" {
+ */
+#endif
+/*
+ * Now, check if XTIMER_HZ is a frequency multiple of 1MHz (or 15675, which is
```
typo: 156**7**5 -> 15625

gebart commented on sys/include/xtimer/tick_conversion.h:
```diff
@@ -129,10 +269,13 @@ inline static uint64_t _xtimer_usec_from_ticks64(uint64_t ticks) {
 #error Unknown hardware timer frequency (XTIMER_HZ), check board.h and/or add an implementation in sys/include/xtimer/tick_conversion.h
 #endif
+/*
+ * End of optimisations
```
I would prefer to have this comment on the line with the `#endif` on L275, and related to the if condition. Suggestion:
```c
#endif /* XTIMER_HZ == xxx */
```

kYc0o commented:
Sorry for forgetting this PR since looong time, I was kind of waiting for big changed on xtimer but apparently that won't come soon.

I want to update and if possible merge this PR to support another frequency for the mega-xplained board.

ZetaR60 commented:
Some new relevant PRs: #9280, #9283

kYc0o commented:
Yes! I think that will be the way to go for supporting other clock sources in xtimer.
